pageserver/src/tenant/tasks.rs

```diff
@@ -29,10 +29,14 @@ pub fn start_background_loops(tenant: &Arc<Tenant>, init_done: Option<&completio
         false,
         {
             let tenant = Arc::clone(tenant);
-            let init_done = init_done.cloned();
+            let background_jobs_can_start = background_jobs_can_start.cloned();
             async move {
-                completion::Barrier::maybe_wait(init_done).await;
-                compaction_loop(tenant)
+                let cancel = task_mgr::shutdown_token();
+                tokio::select! {
+                    _ = cancel.cancelled() => { return Ok(()) },
+                    _ = completion::Barrier::maybe_wait(background_jobs_can_start) => {}
+                };
```

Comment:
NIT: these lines repeat several times, with either "wait" or "maybe_wait",  could this be refactored to common code ?

Reply:
Yes they do repeat ... I don't know of a better abstraction here; what we want is to stop but not with an error.

With `try_join`:

```rust
use futures::future::FutureExt;
use std::future::ready;

let cancel = cancel.cancelled().then(ready(Err(())));
let can_start = completion::Barrier::maybe_wait(...).then(ready(Ok(()));

// bothers me the await is invisible
if tokio::try_join!(cancel, can_start).is_err() {
    return Ok(());
}
```

Of course, it could just be throw into the depths next to `maybe_wait` but I think then we start to drift to problmatic abstractions as in I think it's valuable to have the control flow visible.

In general I dislike use of `ControlFlow` at the level we have it in our codebase, which I think would be too much for this because what we want to do is a single thing.

At least right now I don't think this is a problem, the copypasteable code clearly has no bugs rather than no clear bugs.

```
// I don't think this is much more better.
//
// impl FutureExt {
//     fn race<A, B>(self, other: impl Future<Output = B>) -> RaceFuture<A, B>
//         where Self: Future<Output = A>;
// }
//
// impl<A, B> std::future::Future for RaceFuture<A, B> {
//    type Output = Either<A, B>;
// 
//    ...
// }

if completion::Barrier::maybe_wait(...).race(cancel.cancellation()).await.is_right() {
    return Ok(());
}

// OR:

if completion::Barrier::maybe_wait_cancellably(maybe_barrier, &cancel).await.is_cancel() {
    return Ok(());
}
```

On top of all the variations I can imagine I don't find them much more appealing.